artiq/tools.py

```diff
@@ -143,3 +143,38 @@ def get_user_config_dir():
     dir = user_config_dir("artiq", "m-labs", major)
     os.makedirs(dir, exist_ok=True)
     return dir
+
+
+# AsyncProxyServer simple class to do socket swap/being a proxy of a remote socket
+class AsyncProxyServer(AsyncioServer):
+    _PROXY_BUFFER_LIMIT = 100 * 1024 ** 2
+
+    def __init__(self, remote: (str, int)):
+        AsyncioServer.__init__(self)
+        self.remote = remote
+
+    # TODO: optimize this to be zero-copy based
```

Comment:
Not necessary considering the volume of moninj traffic. 
Also I had suggested decoding the moninj protocol in the master and then making a standard sync_struct+RPC interface; this way there is also only one connection to the (more resource limited) core device per master.

Reply:
can moninj be shared with concurrently-running dashboards? if not, we might need some queuing in master RPC

Reply:
Yes. And all you might need, if anything, is ``allow_parallel=False`` in ``pc_rpc.Server``. Do not make this more complicated than it should be.